### ISSUE TITLE
[BUG] mlDeploy fails stating "Unable to get version of MarkLogic; cause: 401 Unauthorized"

### ISSUE THREAD

**Issue by s3-4v** [marklogic/marklogic-data-hub#5335]
When running `./gradlew mlDeploy -PenvironmentName=int` (the env name is relevant here), I am seeing the following error repeatedly (masking proprietary names):

> Unable to store properties, cause: int-C:\********\********\marklogic\build\ml-javaclient-util\module-timestamps.properties (The filename, directory name, or volume label syntax is incorrect)
> null
> 

I believe the correct construction of that file path should be C:\********\********\marklogic\build\ml-javaclient-util\int-module-timestamps.properties.

This is with Gradle 6.2.2 on Windows 10 Pro 2004,  DHF 5.4, MarkLogic 10.0-6 on CentOS 7.9.2009.

Please let me know if I can help by providing any additional information.

Thank you.

**Comment by rjrudin** [comment]
Thanks @s3-4v - we had an internal user reported that with DHF 5.2.6 (which uses ml-gradle 3.x), and the common part is that Windows is being used. 

Are you modules still getting loaded? That message should just mean that DHF is not able to capture when it loaded modules, but the modules should still get loaded.

**Comment by nic-gibson** [comment]
@rjrudin -  Adding to that, we saw the same issue without the Gradle wrapper on
* Windows 10 Enterprise
* with DHF 5.2.6 
* Gradle 6.3
* ml-gradle 3.17.1
* Java 1.8.0_251 

One of our dev team reported seeing this on a Linux deploy too but I have no details yet.

**Comment by s3-4v** [comment]
Thanks, @rjrudin, well, it's interesting that you should ask that question, because the same client reported, since upgrading to DHF 5.4, that they would run mlLoadModules, which appeared to build successfully, yet the modules themselves wouldn't be updated. 

However, that may be related to another apparent bug which I was planning to create a separate ticket for: mlDeploy fails stating "Unable to get version of MarkLogic; cause: 401 Unauthorized" -- when that happens, running mlLoadModules effectively fails, until one runs tasks like mlDeployDatabases and mlDeployServers and some other deployment tasks seen in mlDeploy --dry-run so that the deployment essentially completes.

Still the problem remains that running mlDeploy ends with mlDeployApp failing with the following:

> Execution failed for task ':mlDeployApp'.
> > Unable to get version of MarkLogic; cause: 401 Unauthorized: [{"errorResponse":   {"statusCode":401,
>      "status":"Unauthorized",
>      "message":"401 Unauthorized"
>     }
>   }]
> 

I tried setting mlServerVersion=10 in gradle.properties, just in case, didn't help.

I am able to authenticate with the Management API using the same credentials just fine, and it runs for 11 minutes, presumably successfully authenticating many times, before this.

Please let me know if I can provide more information, and I'd be happy to report this problem in another bug report if it would be helpful.

Thank you.

**Comment by rjrudin** [comment]
I'll try to reproduce. The only time DHF uses the value of environmentName is when it constructs a module timestamps file - it prepends the environmentName to user-modules-deploy-timestamps.properties, which should of course result in e.g. "int-user-modules-deploy-timestamps.properties". 

I used to develop on Windows (been a few years), and I do recall issues with where -P flags existed on the command line. You could try putting -PenvironmentName before mlDeploy to see if that helps. Also, DHF 5.4.0 uses Gradle 6.4 for its Gradle wrapper version - I don't know if that will help, but given that our Windows tests are using Gradle 6.4, it's worth a try.

**Comment by rjrudin** [comment]
Oh hey - are you setting mlModuleTimestampsPath in gradle.properties ? That could be the issue here if it's an absolute path - that sounds more like a potential ml-gradle bug (though I'd also say - don't use an absolute path).

**Comment by s3-4v** [comment]
@nic-gibson Interesting, because presently the DHF 5.4 upgrade procedure documented at https://docs.marklogic.com/datahub/5.4/upgrade.html states in step 2:

> Update the /your-project-root/gradle/wrapper/gradle-wrapper.properties file.
> 
> Set distributionUrl to https\://services.gradle.org/distributions/gradle-6.2.2-bin.zip.
> 
> distributionUrl=https\://services.gradle.org/distributions/gradle-6.2.2-bin.zip

**Comment by s3-4v** [comment]
@rjrudin Nope, not setting mlModuleTimestampsPath, but thanks for asking and thinking of that, and for trying to reproduce.

As far as I know, it's not causing any problems, but the other bug https://github.com/marklogic/marklogic-data-hub/issues/5336 I just reported may or may not be related, but it's definitely causing some problems for the client.

**Comment by s3-4v** [comment]
Sorry, @nic-gibson, comment https://github.com/marklogic/marklogic-data-hub/issues/5335#issuecomment-794402262 was for @rjrudin (please see that comment if you missed it, thanks)

**Comment by rjrudin** [comment]
@s3-4v I haven't found it yet, but do you have a support ticket open for this? If so, can I close this and #5336?

**Comment by s3-4v** [comment]
@rjrudin I understand. I conflated both bug reports into the one existing support ticket. I was notified they opened an investigation.

Just to be clear, I am seeing this apparent bug both locally and on the client's remote. The customer is getting it on their server, too. However, the other bug (#5336) appears to be localized to their server only. 

So yes I think it's OK to close those issues here. Thank you again for the help.
